File: arrays/prob1160.py
# 1160 Find words can be formed by Characters

class Solution(object):
    def countCharacters(self, words, chars):
        """
        :type words: List[str]
        :type chars: str
        :rtype: int
        """
        d = [0] * 26
        for c in chars:
            d[ ord(c) - ord('a') ] += 1
            
        ans = 0
        for word in words:
            wc = [0] * 26
            for c in word:
                i = ord(c) - ord('a')
                wc[i] += 1
            flag = True
            for i in range(26):
                if wc[i] > d[i]:
                    flag = False
                    break
            if flag:
                ans += len(word)
        return ans
        
        

File: hashtable/prob1160.kt
class Solution {
    fun countCharacters(words: Array<String>, chars: String): Int {
        
        var m = mutableMapOf<Char, Int>()
        for (c in chars) {
            m[c] = m.getOrDefault(c, 0) + 1
        }
    
        val f = fun(word: String, m: MutableMap<Char,Int>): Boolean {
            var m1 = mutableMapOf<Char, Int>()
            for (c in word) {
                m1[c] = m1.getOrDefault(c, 0) + 1
            }
            for ( (k, v) in m1 ) {
                if (v > m.getOrDefault(k, 0)) {
                    return false
                }
            }
            return true
        }
        
        return words.filter({it -> f(it, m)}).fold(0) { acc, it-> acc + it.length}
        
    }

}
